internals/link.py

```diff
@@ -0,0 +1,95 @@
+import logging
```

Comment:
Each of these files should have the copyright header comments.
You can copy them from any of the other python files.
Just make sure to edit the year number to be 2023 in these new files.